### PR TITLE
Timer leaks/Multiple disconnect notifications

### LINKED ISSUE #255: node-nats leaves ping timer out when disconnected triggers a close
stream.on('close performs the logic on whether to fire the close event, but it doesn't clean ping timeout.

``` 
stream.on('close', function(hadError) {
// needs the following added:
        if (client.pingTimer) {
            clearTimeout(client.pingTimer);
            delete client.pingTimer;
        }
```

### PR DESCRIPTION
**Pull request by aricart**
FIX #255, #256, #257.

Currently, when the client is closed, only the ping timer is cleaned up. Other timers such as timeouts on a subscription or a request would remain live.
Fixed so that if the client fails to reconnect and issues a 'close' from the stream, that timeout resources are also cleaned up.

Disconnect notifications are raised by the stream handler whenever close is called on the stream, this is incorrect because the close handler is called even if a connection on the current socket was not established. This makes downstream clients (such as node-nats-streaming) have false notifications that complicate the logic of tests and other things. With this change a 'disconnect' is only notified if the current socket was actually active.

### CODE REVIEW

File: lib/nats.js [hunk 1]
```diff
@@ -831,10 +843,7 @@ Client.prototype.initState = function() {
  * @api public
  */
 Client.prototype.close = function() {
-    if (this.pingTimer) {
-        clearTimeout(this.pingTimer);
-        delete this.pingTimer;
-    }
+    this.cleanupTimers();
```
Comment (aricart) [inline lib/nats.js]: this is not needed

Comment (kozlovic) [reply lib/nats.js]: why?

Comment (aricart) [reply lib/nats.js]: close on the stream will do it.

Comment (kozlovic) [reply lib/nats.js]: Ok, we will still have all the subscriptions by that time, right? That is, there is no place where we would remove them by the time we get to stream.close and therefore are unable to cancel subscriptions' timer?

Comment (aricart) [reply lib/nats.js]: actually - my bad the cleanup is required as closing the stream from `close` will not emit the 'close' event on the stream.

File: lib/nats.js [hunk 2]
```diff
@@ -588,7 +588,19 @@ Client.prototype.connectCB = function() {
     this.flushPending();
 };
 
+/**
+ * @api private
+ */
+Client.prototype.cancelHeartbeat = function() {
+    if (this.pingTimer) {
```
Comment (derekcollison) [inline lib/nats.js]: Should we not set pingTimer to null or undefined in case this gets called again?

File: lib/nats.js [hunk 3]
```diff
@@ -846,6 +853,33 @@ Client.prototype.close = function() {
...
+ */
+Client.prototype.cleanupTimers = function() {
+    this.cancelHeartbeat();
+
+    if(this.respmux && this.respmux.requestMap) {
+        for (const p in this.respmux.requestMap) {
+            if (this.respmux.requestMap.hasOwnProperty(p)) {
+                this.cancelMuxRequest(p);
+            }
+        }
```
Comment (derekcollison) [inline lib/nats.js]: Again should we null or set to undefined in case of being called again?

**Pull request merged**